File: java/Solution543.java
package leetCode.java;

//DFS O(V) O(H)
class Solution543 {
    int res = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        dfs(root);
        return res;
    }

    public int dfs(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int left = dfs(root.left);
        int right = dfs(root.right);
        res = Math.max(res, left + right); // 題目所求是edge邊數, 不是node數
        return Math.max(left, right) + 1;
    }
}

File: c++/Solution543.cpp
//DFS O(V) O(H)
class Solution543 {
public:
    int res = 0;

    int diameterOfBinaryTree(TreeNode* root) {
        //不是要求高度, 而是任意兩點最遠距離
        dfs(root);
        return res;
    }

    int dfs(TreeNode* root) {
        if(!root) return 0;

        int left = dfs(root->left);
        int right = dfs(root->right);
        res = max(res, left + right);//紀錄最遠距離
        return max(left, right) + 1;
    }
};